### ISSUE TITLE
Change sort order of playlist

### ISSUE THREAD

**Issue by t89** [Rigellute/spotify-tui#76]
An option to sort the playlist in different ways would be nice: Show the most recently added songs on top for example.

**Comment by SgiobairOg** [comment]
Completely new to Rust but, as this is a super interesting project could I offer to take a crack at this issue?

**Comment by jfaltis** [comment]
I would suggest to use Shift+S or Alt+s to enter a "sort mode" which basically just listens for the next keypress and processes it to decide to sort by what criteria. E.g t for title, a for artist, l for length etc. Maybe it could also show some kind of "Statusbar" with the sorting criterias and corresponding keys while being in "sort mode"

**Comment by Rigellute** [comment]
Upper case `S` (i.e. `<Shift+S>`) sounds like a good binding to me.

And if you want to take a crack at it @SgiobairOg, be my guest!

**Comment by SgiobairOg** [comment]
So, handling the key inputs is straightforward enough, even with no Rust experience. 

Could someone point me to the code that handles populating the track table though? I'm assuming I can just grab the tracks from `app.track_table.tracks`, and sort them with `slice::sort_by`. Not sure how to update the display and set the current track if one is playing.

**Comment by Debdutto** [comment]
@SgiobairOg Are you still working on this? Was hoping to have a crack at it over the weekend?

**Comment by SgiobairOg** [comment]
I was, but don’t actually know how to do the sort and haven’t had time to figure it out. I can commit a branch if you want so you can have what I’ve got so far.

**Comment by LU15W1R7H** [comment]
I've been spending some time on this issue and there seem to be some obstacles in implementing this...

I was successful in sorting the tracks by Title, Artist, Album and Duration depending on the keyboard input. But the difficulty lies in starting the playback and have the tracks play in the same order as they are in the `TrackTable`. If I'm right I must pass all the `URI`s of all tracks in the correct order to the spotify-api. So far we've just passed `None` as argument, since we just used the default ordering. I hope this doesn't have any noticeable performance implications when playing big playlists.

I will continue working on it.

**Comment by Rigellute** [comment]
You are right @LU15W1R7H, we can sort the display for the tracks and can even get the playback to work in that order - but the limitation is that Spotify will only play those tracks that have been sorted (once you get to the end of that list, playback will stop).

Perhaps this is intentional, but means we will need to correctly manage unsorted (i.e. default) mode, which has the full context of the current playback (playlist/album etc.). That is to say we don't specify an array of URLs.

Totally doable, but more state to manage!

**Comment by LU15W1R7H** [comment]
@Rigellute yeah, thats totally true. I would also prefer another solution, I just haven't found one so far. It seems the spotify-api just doesn't provide the ability the play the songs in some predefined order...
